Check automatically finish setup if allowed

solooboroten opened the pull request:
Under some conditions, the first request to owncloud_autosetup_url,
returns a status other than 200. In this case, all subsequent requests
owncloud_autosetup_url return a status of 200.

Signed-off-by: Aleksey Avdeev <solo@altlinux.ru>

solooboroten commented:
This patch is to fix #39.

drybjed commented:
I'm sorry, but I don't accept `ignore_errors: True` as a solution to bugs. Bugs should be fixed, not ignored.

I think that `debops.owncloud` role might not work properly due to release of ownCloud 9 and related changes. I'll check it when I have some time.

solooboroten commented:
In this case, the problem is that the virtual machine requires too much memory (1548 MB of the available 1978 MB), if the amount of memory required to reduce (eg to the default value of 512 MB, as is done in [finish_setup/noproblem](https://github.com/solooboroten/Vagrantfile/tree/finish_setup/noproblem)) - check "Automatically finish setup if allowed" to work normally.

I think that the mechanism of the problem the following .: The first time the `owncloud_autosetup_url` run some initialization, which can not be completed within the time limit (for example due to the fact that the base system is actively uses the swap disk subsystem or her busy with other bulk operations). Since the initialization process is not completed, and the time limit is exhausted - the server returns a status other than `200`. (If we judge by the return status is `401` - it seems that the server underachieving bridge the database.) All subsequent queries run normally, as all initialization processes (running the first query) have time to be completed by that time.

As in this situation to do without repeated requests, I do not know. But the abandonment of `ignore_errors`, an idea I have.

drybjed commented:
If the `401` status code is also valid, and ownCloud works normally after that, then you should be able to specify multiple return codes in the `uri` Ansible module, like this:

```
- name: Do stuff
  uri:
    url: '{{ owncloud_autosetup_url }}'
    status_code: '200,401'
```

Can you test if that works?

Also, 1500MB+ just to configure an PHP5 application like ownCloud? That seems overkill, unless you are running something else on that VM...

solooboroten commented:
I'll check this version: [4418a47306dc1364c681cf801bb1131feb253ced](https://github.com/solooboroten/ansible-owncloud/commit/4418a47306dc1364c681cf801bb1131feb253ced). The result will be about an hour.

> Also, 1500MB+ just to configure an PHP5 application like ownCloud? That seems overkill, unless you are running something else on that VM...

This requirement of the customer ...

solooboroten commented:
The virtual machine is configured with the commit [4418a47306dc1364c681cf801bb1131feb253ced](https://github.com/solooboroten/ansible-owncloud/commit/4418a47306dc1364c681cf801bb1131feb253ced) (see [finish_setup/fix (e54ee570e35d33693039300db5e269658775df3b)](https://github.com/solooboroten/Vagrantfile/commit/e54ee570e35d33693039300db5e269658775df3b)) was created. To quote the log:

```
$ vagrant up --provider virtualbox
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'ubuntu/trusty32'...
...
==> default: Running provisioner: ansible...
    default: Running ansible-galaxy...
ansible-galaxy install --role-file=/home/solo/Vagrant/tests/debops.owncloud/provisioning/requirements.yml --roles-path=/home/solo/Vagrant/tests/debops.owncloud/provisioning/roles --force
...
- executing: git clone https://github.com/solooboroten/ansible-owncloud.git debops.owncloud
- executing: git archive --prefix=debops.owncloud/ --output=/tmp/.private/solo/tmpXu6AG4.tar 4418a47306dc1364c681cf801bb1131feb253ced
- extracting debops.owncloud to /home/solo/Vagrant/tests/debops.owncloud/provisioning/roles/debops.owncloud
- debops.owncloud was installed successfully
- dependency debops.secret is already installed, skipping.
    default: Running ansible-playbook...
PYTHONUNBUFFERED=1 ANSIBLE_NOCOLOR=true ANSIBLE_HOST_KEY_CHECKING=false ANSIBLE_SSH_ARGS='-o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s' ansible-playbook --connection=ssh --timeout=30 --limit='default' --inventory-file=/home/solo/Vagrant/tests/debops.owncloud/.vagrant/provisioners/ansible/inventory -v provisioning/playbook.yml

PLAY [all] ********************************************************************
...
NOTIFIED: [debops.nginx | Reload nginx] *************************************** 
changed: [default] => {"changed": true, "name": "nginx", "state": "started"}

TASK: [debops.owncloud | Automatically finish setup if allowed] *************** 
ok: [default] => {"cache_control": "no-cache, must-revalidate", "changed": false, "connection": "keep-
alive", "content_length": "43", "content_security_policy": "default-src 'none';script-src 'self' 'unsa
fe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self
';media-src 'self'", "content_type": "application/json; charset=utf-8", "date": "Wed, 06 Apr 2016 20:2
2:31 GMT", "expires": "Thu, 19 Nov 1981 08:52:00 GMT", "json": {"message": "Current user is not logged
 in"}, "pragma": "no-cache", "redirected": false, "server": "nginx", "set_cookie": "occw5e2qp9z4=vpfci
03gpn4cpp28v21ojerjk2; path=/; HttpOnly, oc_sessionPassphrase=BMjx03DsPwVAxBl0VgS7RsXp3UFBanSjvg7p%2BM
1JHRiwf3xmKVrQ%2Bueo9J7ROWLC19rFfffI65Gy2tjxcjEx0R0MrlnmPpAxE4rr277VpIv2t%2FSRiWdYxJ%2BubS2yA3Ku; path
=/; httponly", "status": 401, "x_content_type_options": "nosniff", "x_download_options": "noopen", "x_
frame_options": "Sameorigin", "x_permitted_cross_domain_policies": "none", "x_robots_tag": "none", "x_
xss_protection": "1; mode=block"}

TASK: [debops.owncloud | Check automatically finish setup if allowed] ********* 
ok: [default] => {"_content_encoding": "gzip", "cache_control": "no-store, no-cache, must-revalidate, 
post-check=0, pre-check=0", "changed": false, "connection": "keep-alive", "content_length": "10491", "
content_location": "http://localhost:8087/index.php", "content_security_policy": "default-src 'self'; 
script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src * data: blob:;
 font-src 'self' data:; media-src *; connect-src *", "content_type": "text/html; charset=UTF-8", "date
": "Wed, 06 Apr 2016 20:22:36 GMT", "expires": "Thu, 19 Nov 1981 08:52:00 GMT", "pragma": "no-cache", 
"redirected": false, "server": "nginx", "set_cookie": "occw5e2qp9z4=bb5s672ljlgiusbfj8dffmdbb6; path=/
; HttpOnly, oc_sessionPassphrase=g%2BoPah9W6ozf%2BuPdTNfB%2FULVkVHNtUpHAz3tCJgn7TwHbI3lbZV3eSMNyc8ztG4obQo60mgR4zf%2FoeMeoMEFuKqT11pXsplbEdUABhE3bmgVYQa4ZE%2BW6SXpab3WX0oz; path=/; httponly", "status": 200, "transfer_encoding": "chunked", "x_clacks_overhead": "GNU Terry Pratchett", "x_content_type_options": "nosniff", "x_download_options": "noopen", "x_frame_options": "Sameorigin", "x_permitted_cross_domain_policies": "none", "x_robots_tag": "none", "x_xss_protection": "1; mode=block"}

TASK: [debops.owncloud | Allow to use the PHP CLI even if OCMemcacheAPCu is used] *** 
skipping: [default]
...
PLAY RECAP ******************************************************************** 
default                    : ok=185  changed=97   unreachable=0    failed=0   
```

drybjed commented:
If it's a production environment then no worries, I thought you were just testing. :-)

So the task now passes? Then you don't need the second one. It seems that Travis CI test is broken due to some issue with APT repository... Gonna have to debug that.

drybjed commented:
OK, the test is fixed and passes. When you remove the redundant task, it should be fine to merge.

solooboroten commented:
I think that weaken the final inspection is undesirable. As a sign of success, still has to stay the return code `200`. Otherwise it will be impossible to catch a situation where the return code is always `401`. (I'm not so experienced, but I see no evidence of the impossibility of such a logical state of the system.)

In my opinion there are the following list of choices system state:

The system buildable, if:
1. At the request `owncloud_autosetup_url` returned code `200`.
2. When several consecutive requests `owncloud_autosetup_url` returned multiple codes `401` and the last -- `200`. (My situation is a special case of this variant: the first query returns `401`, the second -- `200`.)

The system is not buildable, if:
1. When several consecutive requests `owncloud_autosetup_url` returned only code `401` (No `200` at the end of the code sequence.)
2. When prompted `owncloud_autosetup_url` returned code other than `200` or `401`.

In my opinion, the number of distinct states of the system with the number of  `owncloud_autosetup_url` requests are related as follows:
1. One request. OK only if the query returns the status `200`. If the status of the `401` considered valid, it will be considered valid and the status code the server always returns only `401`. (That is not right, in my opinion.)
2. Two query. OK if any of the query returns a `200`. Let's return code `401` for the first request. (It is this version is implemented in the current version of the patch.)
3. The sequence of several requests (the number is given by the variable). OK if any of the queries returns `200`. Assume for all queries return code `401` except the last.

solooboroten commented:
At commit [8d3dfb23604318209c0ab0c96cd7430fb9c1b9e8](https://github.com/solooboroten/ansible-owncloud/commit/8d3dfb23604318209c0ab0c96cd7430fb9c1b9e8) to change task names.

solooboroten commented:
> 1. The sequence of several requests (the number is given by the variable). OK if any of the queries returns `200`. Assume for all queries return code `401` except the last.

This option does not seem to realize, without creating additional roles. Somehow, without additional roles to check the result of the previous iteration in the loop. (`Include` in a loop does not work, at least in ansible-1.9.2.)

drybjed commented:
OK, I agree that you are right about status `401` being not enough to determine if role finished successfully. I suppose that executing the `uri` module twice should suffice for the test.

Right now Ansible devel is used for testing, because of that the test fails. I'll switch it to the stable ranch ASAP, run the test again and merge when it's OK.

solooboroten commented:
The commit [8d3dfb23604318209c0ab0c96cd7430fb9c1b9e8](https://github.com/solooboroten/ansible-owncloud/commit/8d3dfb23604318209c0ab0c96cd7430fb9c1b9e8) bug. Fixed to commit [de271bee1997149ab72e7824a33cb3cbfd599689](https://github.com/solooboroten/ansible-owncloud/commit/de271bee1997149ab72e7824a33cb3cbfd599689). To quote the log (see [finish_setup/fix (3d9f04962e3c87d77496a03657ce2bc846319152)](https://github.com/solooboroten/Vagrantfile/commit/3d9f04962e3c87d77496a03657ce2bc846319152)):

```
$ vagrant up --provider virtualbox
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'ubuntu/trusty32'...
...
    default: Running ansible-galaxy...
ansible-galaxy install --role-file=/home/solo/Vagrant/tests/debops.owncloud/provisioning/requirements.yml --roles-path=/home/solo/Vagrant/tests/debops.owncloud/provisioning/roles --force
...
- executing: git clone https://github.com/solooboroten/ansible-owncloud.git debops.owncloud
- executing: git archive --prefix=debops.owncloud/ --output=/tmp/.private/solo/tmpUe0ulS.tar de271bee1997149ab72e7824a33cb3cbfd599689
- extracting debops.owncloud to /home/solo/Vagrant/tests/debops.owncloud/provisioning/roles/debops.owncloud
- debops.owncloud was installed successfully
- dependency debops.secret is already installed, skipping.
    default: Running ansible-playbook...
PYTHONUNBUFFERED=1 ANSIBLE_NOCOLOR=true ANSIBLE_HOST_KEY_CHECKING=false ANSIBLE_SSH_ARGS='-o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ControlMaster=auto -o ControlPersist=60s' ansible-playbook --connection=ssh --timeout=30 --limit='default' --inventory-file=/home/solo/Vagrant/tests/debops.owncloud/.vagrant/provisioners/ansible/inventory -v provisioning/playbook.yml

PLAY [all] ******************************************************************** 
...
NOTIFIED: [debops.nginx | Reload nginx] *************************************** 
changed: [default] => {"changed": true, "name": "nginx", "state": "started"}

TASK: [debops.owncloud | Test availability of automatically finish setup] ***** 
ok: [default] => {"cache_control": "no-cache, must-revalidate", "changed": false, "connection": "keep-
alive", "content_length": "43", "content_security_policy": "default-src 'none';script-src 'self' 'unsa
fe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self
';media-src 'self'", "content_type": "application/json; charset=utf-8", "date": "Sat, 09 Apr 2016 09:0
4:58 GMT", "expires": "Thu, 19 Nov 1981 08:52:00 GMT", "json": {"message": "Current user is not logged
 in"}, "pragma": "no-cache", "redirected": false, "server": "nginx", "set_cookie": "oc6njsz9toh5=7uj1j
vqmsbqa76tdgn0h34pc77; path=/; HttpOnly, oc_sessionPassphrase=DbnUXwxkvRdp8OsyO4KB2dztA0rQto1jqeU50ctP
aGU3cfCl2UrBJ3%2Bnl8XUig8tKxguYHThgavOntPgRHyC1kbgoYySI57KFRDdpwbs45HcBpYDwS94pbrHbAH5Trip; path=/; ht
tponly", "status": 401, "x_content_type_options": "nosniff", "x_download_options": "noopen", "x_frame_
options": "Sameorigin", "x_permitted_cross_domain_policies": "none", "x_robots_tag": "none", "x_xss_pr
otection": "1; mode=block"}

TASK: [debops.owncloud | Automatically finish setup if allowed] *************** 
ok: [default] => {"_content_encoding": "gzip", "cache_control": "no-store, no-cache, must-revalidate, 
post-check=0, pre-check=0", "changed": false, "connection": "keep-alive", "content_length": "10491", "
content_location": "http://localhost:8087/index.php", "content_security_policy": "default-src 'self'; 
script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src * data: blob:;
 font-src 'self' data:; media-src *; connect-src *", "content_type": "text/html; charset=UTF-8", "date
": "Sat, 09 Apr 2016 09:05:03 GMT", "expires": "Thu, 19 Nov 1981 08:52:00 GMT", "pragma": "no-cache", 
"redirected": false, "server": "nginx", "set_cookie": "oc6njsz9toh5=34kh4kfoq6jtmasbtdp3g09591; path=/
; HttpOnly, oc_sessionPassphrase=1nBtuIY39%2FHsoiijD4k2uWEvlzQd7sGm3shTzJtAbRQb5HsMG9z%2BEHnylJQsJd3cxu%2FY%2Fo%2BHLhjiMFO6b4zfLrbvfn3FBjcTXPG2fOCeol9S6kMaEY%2Bb8h1e64%2FsVPk%2B; path=/; httponly", "status": 200, "transfer_encoding": "chunked", "x_clacks_overhead": "GNU Terry Pratchett", "x_content_type_options": "nosniff", "x_download_options": "noopen", "x_frame_options": "Sameorigin", "x_permitted_cross_domain_policies": "none", "x_robots_tag": "none", "x_xss_protection": "1; mode=block"}
...

PLAY RECAP ******************************************************************** 
default                    : ok=185  changed=97   unreachable=0    failed=0   
```

ypid commented:
I just implemented a more reliable way to setup ownCloud using `occ maintenance:install`. I will open a PR when I tested this a bit more.
